src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -93,13 +102,168 @@ Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
+    RAY_LOG(INFO) << "Work stealing is not enabled, so we return the worker "
+                  << thief_addr.worker_id << " without stealing";
+    ReturnWorker(thief_addr, was_error);
+    return;
+  }
+
+  RAY_LOG(INFO) << "Stealing work now! Thief is worker: " << thief_addr.worker_id;
+
+  // Look for a suitable victim
+  absl::flat_hash_map<rpc::WorkerAddress, LeaseEntry>::iterator victim =
+      worker_to_lease_entry_.begin();
+  for (absl::flat_hash_map<rpc::WorkerAddress, LeaseEntry>::iterator candidate =
+           worker_to_lease_entry_.begin();
+       candidate != worker_to_lease_entry_.end(); ++candidate) {
+    RAY_LOG(DEBUG) << "Current victim: " << victim->first.worker_id << " with "
+                   << victim->second.tasks_in_flight_ << " tasks in flight";
+    RAY_LOG(DEBUG) << "Current candidate: " << candidate->first.worker_id << " with "
+                   << candidate->second.tasks_in_flight_ << " tasks in flight";
+
+    if (candidate->second.current_scheduling_key_ != scheduling_key) {
+      continue;
+    }
+    if (victim->second.current_scheduling_key_ != scheduling_key ||
+        victim->first.worker_id == thief_addr.worker_id ||
+        ((candidate->second.stealable_tasks_.size() >
+          victim->second.stealable_tasks_.size()) &&
+         candidate->first.worker_id != thief_addr.worker_id)) {
+      victim = candidate;
+    }
+  }
+
+  // Check that the victim is a suitable one
+  if (victim->second.current_scheduling_key_ != scheduling_key ||
+      victim->first.worker_id == thief_addr.worker_id) {
+    RAY_LOG(INFO) << "Could not find a suitable victim for stealing! Returning worker "
+                  << thief_addr.worker_id;
+
+    ReturnWorker(thief_addr, was_error);
+    return;
+  }
+
+  RAY_CHECK(victim->second.current_scheduling_key_ == scheduling_key);
+  RAY_CHECK(victim->first.worker_id != thief_addr.worker_id);
+
+  RAY_LOG(DEBUG) << "Victim is worker " << victim->first.worker_id << " and has "
+                 << victim->second.tasks_in_flight_ << " tasks in flight.";
+
+  // Make sure that the victim has more than 1 task in flight
+  RAY_LOG(DEBUG) << "victim->second.tasks_in_flight_: " << victim->second.tasks_in_flight_
+                 << " victim->second.stealable_tasks_.size(): "
+                 << victim->second.stealable_tasks_.size();
+  RAY_CHECK(victim->second.tasks_in_flight_ >= victim->second.stealable_tasks_.size());
+
+  if (victim->second.stealable_tasks_.size() <= 1) {
+    RAY_LOG(DEBUG)
+        << "The designated victim had <= 1 tasks in flight, so we don't steal.";
+
+    ReturnWorker(thief_addr, was_error);
+    return;
+  }
+
+  auto request = std::unique_ptr<rpc::StealWorkRequest>(new rpc::StealWorkRequest);
+  RAY_LOG(DEBUG)
+      << "Calling Steal Work RPC! Maximum number of tasks requested (to steal) = "
+      << victim->second.tasks_in_flight_;
+  request->set_max_tasks_to_steal(victim->second.tasks_in_flight_);
```

Comment:
This flag seems useful in theory since stealing can be expensive. However, I don't think it makes sense to set it to the total number of tasks in flight to the victim. The victim should already be basically limited to that number since its queue size should be <= the tasks in flight, right? (Assuming the owner doesn't send many more tasks to the victim in between this point and when the victim receives the steal request)